astropy/units/quantity.py

```diff
@@ -286,6 +314,98 @@ class Quantity(np.ndarray):
 
     __array_priority__ = 10000
 
+    def __init_subclass__(cls, **kwargs):
```

Comment:
So it's a little complicated.
We want an automatic registry to properly return bound type variables
https://realpython.com/python-type-checking/#returning-self-or-cls

I think in principle we could push this to another PR, but ideally in a future PR we could make the following example return an Angle instance instead of a Quantity and could properly detect if the unit-type was incompatible with the class.

```python
@u.quantity_input
def test(x: u.Quantity["length"]) -> coord.Angle[u.rad]:
    return x * u.rad / u.m

test(2 * u.km)
```

Including this now adds the necessary architecture and doesn't change any of the current functionality in ``quantity_input``. 
I agree that it's odd to do ``SpecificTypeQuantity[u.m]`` over ``Distance[u.m]``, but it is actually meaningful in that it  disallows a raw ``Quantity``

Reply:
Out of scope : Part of the larger conversation on how astropy will want to implement ``typing`` is whether we want to store these in a private dictionary versus in an accessible location, like ``astropy/units/type_hints.py``.